src/service/resize-observer.js

```diff
@@ -0,0 +1,167 @@
+ * instantiate their own resize observers. But currently there are some
+ * performance issues and pooling is desired. For more info, see
+ * https://bugs.chromium.org/p/chromium/issues/detail?id=1157544
+ *
+ * @implements {../service.Disposable}
+ */
+export class ResizeObserverImpl {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+
+    /** @private @const {!ResizeObserver} */
+    this.observer_ = new win.ResizeObserver((e) => this.processEntries_(e));
+
+    /** @private {?WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+    this.lastSizes_ = new WeakMap();
+
+    /** @private {?WeakMap<!Element, !ObserverCallbackDef>} */
+    this.callbacks_ = new WeakMap();
+
+    /** @private @const */
+    this.microtask_ = Promise.resolve();
+  }
+
+  /** @override */
+  dispose() {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    this.callbacks_ = null;
+    this.lastSizes_ = null;
+    this.observer_.dispose();
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!ObserverCallbackDef} callback
+   */
+  observe(element, callback) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    let callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      callbacks = [];
+      this.callbacks_.set(element, callbacks);
+      this.observer_.observe(element);
+    }
+    if (!pushIfNotExist(callbacks, callback)) {
+      return;
+    }
+    const lastSize = this.lastSizes_.get(element);
+    if (lastSize) {
+      this.microtask_.then(() => callCallbackNoInline(callback, lastSize));
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!ObserverCallbackDef} callback
+   */
+  unobserve(element, callback) {
+    if (!RESIZE_OBSERVER) {
+      return;
+    }
+    const callbacks = this.callbacks_.get(element);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.callbacks_.delete(element);
+      this.lastSizes_.delete(element);
+      this.observer_.unobserve(element);
+    }
+  }
+
+  /**
+   * @param {!Array<!ResizeObserverEntry>} entries
+   * @private
+   */
+  processEntries_(entries) {
+    for (let i = entries.length - 1; i >= 0; i--) {
+      const {target, clientRect} = entries[i];
+      let seenTarget = false;
+      for (let j = i + 1; j < entries.length; j++) {
+        if (entries[j] == target) {
+          seenTarget = true;
+          break;
+        }
+      }
+      if (seenTarget) {
+        continue;
+      }
+      const callbacks = this.callbacks_.get(target);
+      if (!callbacks) {
+        continue;
+      }
+      const {width, height} = clientRect;
+      /** @type {../layout-rect.LayoutSizeDef} */
+      const size = {width, height};
+      this.lastSizes_.set(target, size);
+      for (let k = 0; k < callbacks.length; k++) {
+        callCallbackNoInline(callbacks[k], size);
```

Comment:
nit: this is cleaner imo if you pass in the whole thunk. E.g. 
```js
callNoInline(() => callbacks[k](size))
```

Reply:
I was trying to reduce number of function allocations.